internal/config/auth_config_test.go

```diff
@@ -380,6 +425,223 @@ func TestLogoutIgnoresErrorsFromConfigAndKeyring(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestSwitchUserMakesSecureTokenActive(t *testing.T) {
+	// Given we have a user with a secure token
+	authCfg := newTestAuthConfig(t)
+	_, err := authCfg.Login("github.com", "test-user-1", "test-token-1", "ssh", true)
+	require.NoError(t, err)
+	_, err = authCfg.Login("github.com", "test-user-2", "test-token-2", "ssh", true)
+	require.NoError(t, err)
+
+	// When we switch to that user
+	require.NoError(t, authCfg.SwitchUser("github.com", "test-user-1"))
+
+	// Their secure token is now active
+	token, err := authCfg.TokenFromKeyring("github.com")
+	require.NoError(t, err)
+	require.Equal(t, "test-token-1", token)
+}
+
+func TestSwitchUserMakesInsecureTokenActive(t *testing.T) {
+	// Given we have a user with an insecure token
+	authCfg := newTestAuthConfig(t)
+	_, err := authCfg.Login("github.com", "test-user-1", "test-token-1", "ssh", false)
+	require.NoError(t, err)
+	_, err = authCfg.Login("github.com", "test-user-2", "test-token-2", "ssh", false)
+	require.NoError(t, err)
+
+	// When we switch to that user
+	require.NoError(t, authCfg.SwitchUser("github.com", "test-user-1"))
+
+	// Their insecure token is now active
+	token, source := authCfg.ActiveToken("github.com")
+	require.Equal(t, "test-token-1", token)
+	require.Equal(t, oauthTokenKey, source)
+}
+
+func TestSwitchUserUpdatesTheActiveUser(t *testing.T) {
+	// Given we have two users logged into a host
+	authCfg := newTestAuthConfig(t)
+	_, err := authCfg.Login("github.com", "test-user-1", "test-token-1", "ssh", false)
+	require.NoError(t, err)
+	_, err = authCfg.Login("github.com", "test-user-2", "test-token-2", "ssh", false)
+	require.NoError(t, err)
+
+	// When we switch to the other user
+	require.NoError(t, authCfg.SwitchUser("github.com", "test-user-1"))
+
+	// Then the active user is updated
+	activeUser, err := authCfg.ActiveUser("github.com")
+	require.NoError(t, err)
+	require.Equal(t, "test-user-1", activeUser)
+}
+
+func TestSwitchUserErrorsImmediatelyIfTheActiveTokenComesFromEnvironment(t *testing.T) {
```

Comment:
So firstly, this shouldn't really ever be hit because the `Switch` command [protects against](https://github.com/cli/cli/blob/wm/multi-account-ux-rebase/pkg/cmd/auth/switch/switch.go#L157) this already but I wanted to be defensive here in case it happens in future.

The `switch` command here acts the same as it does for `login` and `logout` which have exactly the same check. The idea is that if your token for a host is coming from the environment then we should be pretty suspicious of the user trying to make changes to their authorization set up because there's a good chance they are doing something surprising e.g. expecting that after logging in they are acting as someone else but it turns out it's coming from the env var.

I could go either way on this behaviour across all the commands but since it was in the others it felt sensible to include here.